### ISSUE TITLE
Implement further color correction post processing effects

### ISSUE THREAD

**Issue by jcostello** [godotengine/godot-proposals#4731]
### Describe the project you are working on

3D environment lighting in Godot

### Describe the problem or limitation you are having in your project

Working on 3D environments, specially on the lighting I realize that setting the color mood is not that easy in Godot. Currently Godot's color correction post processing effect is not enough and is not easy to set up the right tones. 

The only thing to have in mind is that in these before and after, I wouldn't take advantage of the indirect light energy when baking light compared with the other engine bellow

Before

![image](https://user-images.githubusercontent.com/1168582/175322257-552e2b46-0f7e-4493-9ed3-871710f12003.png)

After

![image](https://user-images.githubusercontent.com/1168582/175322137-b03303ea-8c8d-4f58-a314-dad3ba3ce31d.png)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be ideal to have shadow/midtones/highlights as well as lift/gamma/gain color correction effects to help setting the right color mood in the scene. Also Having an intensity slider to animate them would be ideal.

Color correction done in other engine. 

Before
![image](https://user-images.githubusercontent.com/1168582/175319491-68ebd96d-4608-4da7-ba83-9531ea4b437b.png)

After
![image](https://user-images.githubusercontent.com/1168582/175319604-667c0f24-a90f-4865-ac8f-5d220dd03ebe.png)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Ideally it would be on the Environment Node

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Could be a shader but IMO it would be nice to have it on the engine available for everyone as it is something that most of the people would tweak

### Is there a reason why this should be core and not an add-on in the asset library?

Same as above

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot-proposals/issues/2196.

All the color correction you've mentioned can already be done using a 3D LUT texture, but I agree exposing built-in shadows/midtones/highlights adjustments would be useful. However, this needs to be done in a way that doesn't cause a performance penalty when you don't use the feature.

**Comment by jcostello** [comment]
@Calinou I agreed. Althought it can be done, it need to be crafted outside and could be a bit tedious. Also I dont know if you can achive the same result when pushing the shadows up or the highlights down. 

Either way, as you said, performance should not be affected when this is not used.

**Comment by VivianVerdant** [comment]
This would be very useful. I use color correction suites in video editing software all the time, and being able to fine tune the way lighting and shadows looks can be powerful in setting the mood and readability in a scene.

**Comment by fire** [comment]
Feel free to salvage my tiny color management system pr.

https://github.com/fire/godot/tree/littlecms

**Comment by QbieShay** [comment]
> Related to #2196.
> 
> All the color correction you've mentioned can already be done using a well-crafted 3D LUT texture, but I agree that exposing built-in shadows/midtones/highlights adjustments would be useful, especially when using ACES tonemapping (as it tends to make things dark). However, this needs to be done in a way that doesn't cause a performance penalty when you don't use the feature.

I think we could use a LUT under the hood and perhaps give a good in-engine tool to generate one?

**Comment by Calinou** [comment]
> I think we could use a LUT under the hood and perhaps give a good in-engine tool to generate one?

LUT generation in the engine is an interesting idea, but it also prevents doing real-time manipulation. I'm not sure how often people would need to change shadows/midtones/highlights during gameplay though.

**Comment by VivianVerdant** [comment]
Being able to make changes in real time, or blend between custom presets would be useful in instances like traveling between indoors and outdoors, or during day-night cycles.

**Comment by QbieShay** [comment]
Yes I'd lean towards blending multiple LUTs, yes :) although this could be more expensive (computation wise) than just regulating a uniform across all shadow passes? maybe we can expose these parameters in project settings or visual server (shadow tint, light tint, etc.)

**Comment by Calinou** [comment]
> although this could be more expensive (computation wise) than just regulating a uniform across all shadow passes? maybe we can expose these parameters in project settings or world environment (shadow tint, light tint, etc.)

Shadows/Midtones/Highlights don't actually affect shadow or light appearance. They affect specific levels of the image's brightness by adjusting parts of the luminance curve. It's kind of like using a Levels filter in an image editor.

**Comment by QbieShay** [comment]
If the agreed approach is to generate LUTs, then i think we need to re-answer the question "can it be done with an addon"

**Comment by jcostello** [comment]
Sure It can, but probably tricky done by an addon. In my opinion this could be in the core as an upgrade of the existing post effect

**Comment by fire** [comment]
The design is that we give icc color profiles or alternative tables to https://github.com/fire/godot/tree/littlecms and it'll generate a lookup table texture by taking a standard color 2d array / texture atlas and multiplying the lut with the texture and then set as the color grade option in godot 4.

**Comment by ArseniyMirniy** [comment]
1. Using LUTs doesn't make it easier for casual audience
2. For now, you can use my Visual Shader for color correction and screen effects: https://github.com/ArseniyMirniy/Godot-4-Free-Color-Correction-and-Screen-Effects-Visual-Shader — it's fairly easy to use and it works well with basic systems (like auto exposure). For 4.3 it's better to use ACES, for 4.4 AgX will probably make more sense.
